tests/formats/smi.cpp

```diff
@@ -0,0 +1,296 @@
+        CHECK((bonds[0][0] == 0 && bonds[0][1] == 1));
+        CHECK((bonds[1][0] == 0 && bonds[1][1] == 3));
+        CHECK((bonds[2][0] == 1 && bonds[2][1] == 2));
+        CHECK((bonds[3][0] == 2 && bonds[3][1] == 3));
+        CHECK((bonds[4][0] == 2 && bonds[4][1] == 5));
+        CHECK((bonds[5][0] == 3 && bonds[5][1] == 4));
+        CHECK((bonds[6][0] == 4 && bonds[6][1] == 5));
+
+        // [CH2+]C[CH+2]
+        frame = file.read_step(6);
+        CHECK(frame[0].get<Property::DOUBLE>("hydrogen_count") == 2.0);
+        CHECK(frame[0].charge() == 1.0);
+        CHECK(frame[2].get<Property::DOUBLE>("hydrogen_count") == 1.0);
+        CHECK(frame[2].charge() == 2.0);
+
+        // C1CC=1
+        frame = file.read_step(8);
+        auto bond_orders = frame.topology().bond_orders();
+        CHECK(bond_orders[0] == Bond::SINGLE);
+        CHECK(bond_orders[1] == Bond::DOUBLE);
+
+        // C=1CC1
+        frame = file.read_step(9);
+        bond_orders = frame.topology().bond_orders();
+        CHECK(bond_orders[0] == Bond::SINGLE);
+        CHECK(bond_orders[1] == Bond::DOUBLE);
+    }
+
+    SECTION("Chirality") {
+        Trajectory file("data/smi/chiral.smi");
+
+        auto frame = file.read();
+        CHECK(frame[1].get("chirality")->as_string() == "CCW TB1");
+        frame = file.read();
+        CHECK(frame[1].get("chirality")->as_string() == "CCW TB15");
+        frame = file.read();
+        CHECK(frame[1].get("chirality")->as_string() == "CW");
+        frame = file.read();
+        CHECK(frame[1].get("chirality")->as_string() == "CCW OH15");
+        frame = file.read();
+        CHECK(frame[1].get("chirality")->as_string() == "CW");
+        frame = file.read();
+        CHECK(frame[1].get("chirality")->as_string() == "CCW");
+    }
+
+    SECTION("Other tests") {
+        Trajectory file("data/smi/test.smi");
+        auto frame = file.read();
+        frame = file.read();
+        CHECK(frame[0].get("is_aromatic")->as_bool());
+        CHECK(frame.get("name")->as_string() == "Benzene");
+
+        while (!file.done()) {
+            frame = file.read();
+        }
+    }
+}
+
+// To use in loops in order to iterate over files in a specific directory.
+struct directory_files_iterator {
+    typedef fs::recursive_directory_iterator iterator;
+    directory_files_iterator(fs::path p) : p_(p) {}
+
+    iterator begin() { return fs::recursive_directory_iterator(p_); }
+    iterator end() { return fs::recursive_directory_iterator(); }
+
+    fs::path p_;
+};
+
+TEST_CASE("Errors in SMI format") {
+    for (auto entry : directory_files_iterator("data/smi/bad/")) {
+        auto test = [=](){
+            // We can throw either when creating the trajectory, or when reading
+            // the frame, depending on the type of error
+            auto file = Trajectory(entry.path().string());
+            file.read();
+        };
+        CHECK_THROWS(test());
+    }
+}
+
+TEST_CASE("Write SMI File") {
```

Comment:
What is the behavior of the writer if there are multiple molecules in the frame? For example, how is `H-O-H       H-O-H` written? We may want a test for this case.

Reply:
That would be `O.O.O`, I've added a test case.